Source/Peripheral.swift

```diff
@@ -127,29 +127,35 @@ public class Peripheral {
 	 Immediately after that `.Complete` is emitted.
 	 */
 	public func discoverServices(serviceUUIDs: [CBUUID]?) -> Observable<[Service]> {
-		let observable = peripheral.rx_didDiscoverServices
+        if let identifiers = serviceUUIDs, let services = self.services?.filter({ identifiers.contains($0.UUID) })
+                                           where identifiers.count == services.count {
+            return ensureValidPeripheralState(Observable.just(services))
+        } else {
+            let observable = peripheral.rx_didDiscoverServices
 			.flatMap {
-				(services, error) -> Observable<[Service]> in
-				if let discoveredServices = services {
-					let mapped = discoveredServices.map { Service(peripheral: self, service: $0) }
-					guard let identifiers = serviceUUIDs else { return Observable.just(mapped) }
-					let uuids = discoveredServices.map { $0.uuid }
-					if Set(uuids) == Set(identifiers) {
-						return Observable.just(mapped)
+				(temp, error) -> Observable<[Service]> in
```

Comment:
`.flatMap { (temp, error) -> Observable<[Service]> in`
don't need to break line here